k2/python/k2/fsa.py

```diff
@@ -70,23 +117,78 @@ def __setattr__(self, name: str, value: Any) -> None:
             object.__setattr__(self, name, value)
         elif isinstance(value, torch.Tensor):
             assert value.shape[0] == self.arcs.values().shape[0]
-            # TODO(fangjun): do we really need a copy here?
+            if name == 'labels':
+                assert value.dtype == torch.int32
+                self.arcs.values()[:, 2] = value
+                return
+
             self._tensor_attr[name] = value.clone()
+
             if name == 'score':
                 assert value.dtype == torch.float32
+                # NOTE: we **reinterpret** the float patterns
+                # to integer patterns here.
                 self.arcs.values()[:, -1] = _as_int(value.detach())
         else:
-            # TODO(fangjun): do we need to copy value?
             self._non_tensor_attr[name] = value
 
     def __getattr__(self, name: str) -> Any:
-        if name in self._tensor_attr:
+        if name == 'labels':
```

Comment:
I'm not a big Python expert, but wouldn't this kind of thing normally be done via a property?  (I.e. the labels and scores getters/setters?)

Reply:
I just tried to add `@property` and `@labels.setter` for reading and writing `labels`. But it seems
that `__setatrr__` has higher precedence than `@labels.setter`. That is, the `setter` of `labels` is not called. 

There is a question on stackoverflow about `properties` and `__setattr__` in Python:
https://stackoverflow.com/questions/15750522/class-properties-and-setattr

Not sure whether `properties` are preferred over `__setattr__` and `__getattr__` or not.